Bugsnag/Breadcrumbs/BugsnagBreadcrumbs.m

```diff
@@ -55,35 +65,29 @@ - (void)addBreadcrumb:(NSString *)breadcrumbMessage {
 }
 
 - (void)addBreadcrumbWithBlock:(BSGBreadcrumbConfiguration)block {
-    if (self.config.maxBreadcrumbs == 0) {
+    if (self.maxBreadcrumbs == 0) {
         return;
     }
     BugsnagBreadcrumb *crumb = [BugsnagBreadcrumb breadcrumbWithBlock:block];
-    if (crumb != nil && [self shouldSendBreadcrumb:crumb]) {
-        dispatch_barrier_sync(self.readWriteQueue, ^{
-            if ((self.breadcrumbs.count > 0) &&
-                (self.breadcrumbs.count == self.config.maxBreadcrumbs)) {
-                [self.breadcrumbs removeObjectAtIndex:0];
-            }
-            [self.breadcrumbs addObject:crumb];
-            // Serialize crumbs to disk inside barrier to avoid simultaneous
-            // access to the file
-            if (self.cachePath != nil) {
-                static NSString *const arrayKeyPath = @"objectValue";
-                NSArray *items = [self.breadcrumbs valueForKeyPath:arrayKeyPath];
-                if ([BSGJSONSerialization isValidJSONObject:items]) {
-                    NSError *error = nil;
-                    NSData *data = [BSGJSONSerialization dataWithJSONObject:items
-                                                                   options:0
-                                                                     error:&error];
-                    [data writeToFile:self.cachePath atomically:NO];
-                    if (error != nil) {
-                        bsg_log_err(@"Failed to write breadcrumbs to disk: %@", error);
-                    }
-                }
-            }
-        });
+    if (!crumb || ![self shouldSendBreadcrumb:crumb]) {
+        return;
     }
+    NSData *data = [self dataForBreadcrumb:crumb];
+    if (!data) {
+        return;
+    }
+    NSUInteger fileNumber;
+    @synchronized (self) {
+        NSMutableArray<BugsnagBreadcrumb *> *breadcrumbs = [self.breadcrumbs mutableCopy];
+        if (breadcrumbs.count >= self.maxBreadcrumbs) {
+            [breadcrumbs removeObjectAtIndex:0];
+        }
+        [breadcrumbs addObject:crumb];
+        self.breadcrumbs = [NSArray arrayWithArray:breadcrumbs];
+        fileNumber = self.lastFileNumber + 1;
+        self.lastFileNumber = fileNumber;
```

Comment:
We could remove the synchronized part here if self.breadcrumbs were atomic and self.lastFileNumber were an atomic int.

Reply:
I did consider that but opted for this approach instead because
* the `OSAtomic` routines are deprecated in favour of C++ `std::atomic` and wanted to avoid switching this to a `.mm`
* the `breadcrumbs` setter is now atomic but that doesn't prevent a race condition where a breadcrumb could be lost if two threads attempt to add one simultaneously - it's the "transaction" that need to be atomic, and therefore requires the `@synchronized` block